cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/functional>
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+using pair_type = thrust::pair<T, T>;
+
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved. See
+ * G. E. Blelloch. Prefix sums and their applications. Technical Report
+ * CMU-CS-90-190, Nov. 1990. S. 1.4
+ * for details
+ */
+template <typename T>
+class recurrence_functor {
+ public:
+  __device__ pair_type<T> operator()(pair_type<T> ci, pair_type<T> cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+template <typename T>
+class ewma_functor_base {
+ public:
+  T beta;
+  ewma_functor_base(T beta) : beta{beta} {}
+
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
+  using ewma_functor_base<T>::ewma_functor_base;
+
+ public:
+  using tupletype = std::conditional_t<nulls, thrust::tuple<bool, int, T>, T>;
+
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    if constexpr (nulls) {
+      bool const valid = thrust::get<0>(data);
+      int const exp    = thrust::get<1>(data);
+      T const input    = thrust::get<2>(data);
```

Comment:
We can unpack these with a structured binding now I think (works in newer CCCL versions).
```suggestion
      auto const [valid, exp, input] = data;
```

Reply:
This seems to conflict later on with the way the lambda below is expressed:

```
...Structured binding cannot be captured...
```

Reply:
Ah yeah this is an unfortunate limitation pre-C++20 (search for "lambda" in https://en.cppreference.com/w/cpp/language/structured_binding). We can work around that with an alias.